### DISCUSSION TITLE
How to do batch updates on data objects?

### DISCUSSION THREAD

**Discussion by VincentBibauw** [pimcore/pimcore#13991] [Q&A / Support]
Hi

I have a data set of +100 000 objects which are synced with an external datasource via the PHP API.
To make the scripts more performant a want to do batch updates on data objects.

Eg. unpublish all the products of a certain category. 

I know I can get all the products and iterate to set them as unpublished, but I'm looking how to do this in 1 query/step.

**Comment by SamyMP** [comment]
I'm not sure how you would do that without iterating on DataObject. You could use the Admin UI, list all the object you want to unpublish, then select "batch edit all" on the "publish" field : 

Total objects "Car" on demo : 
![image](https://user-images.githubusercontent.com/35225430/211259707-0e550be5-96ec-4cd6-a604-481ba043bd09.png)

Filter on a field eg. Car Category on "City Car". Total result is : 
![image](https://user-images.githubusercontent.com/35225430/211260053-9d43ced9-6b30-4a8a-b8d7-6cabc426c320.png)

"Batch edit all" will only apply to the 9 filtered results
![image](https://user-images.githubusercontent.com/35225430/211260280-e22c6984-365b-41f7-8146-477338df7677.png)

**Reply by VincentBibauw** [reply]
Thanks for the reply.

The sync is with an external datasource, so changing in the UI is nog an option.

**Reply by SamyMP** [reply]
Then maybe what you're looking for is an EventListener ? You could trigger it when your objects are imported, and do what you want with the object ?

I think [preAdd](https://github.com/pimcore/pimcore/blob/10.5/lib/Event/DataObjectEvents.php#L25) is what is the most suitable in your case. When the objects are imported, your EventListener could check some conditions (eg. the category of the product) and decide to unpublish the object based on that. There are of course other events, you can find them [in the doc](https://pimcore.com/docs/pimcore/current/Development_Documentation/Extending_Pimcore/Event_API_and_Event_Manager.html)

Is it what you were looking for ?

**Reply by VincentBibauw** [reply]
No, this isn't what I'm looking for. 

When I sync and need to unpublish, I don't necessary get the product for the external data source. So event listeners are not a solution.

**Reply by NiklasBr** [reply]
Are you maybe looking for something like an Entity Manger, for example #10651?

**Reply by VincentBibauw** [reply]
#10651 woud definitely be an improvement for this
